In [0]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import math
import statistics 

In [0]:
import nltk
import re
from nltk.corpus import stopwords

In [0]:
text= '''Those Who Are Resilient Stay In The Game Longer
“On the mountains of truth you can never climb in vain: either you will reach a point higher up today, or you will be training your powers so that you will be able to climb higher tomorrow.” — Friedrich Nietzsche
Challenges and setbacks are not meant to defeat you, but promote you. However, I realise after many years of defeats, it can crush your spirit and it is easier to give up than risk further setbacks and disappointments. Have you experienced this before? To be honest, I don’t have the answers. I can’t tell you what the right course of action is; only you will know. However, it’s important not to be discouraged by failure when pursuing a goal or a dream, since failure itself means different things to different people. To a person with a Fixed Mindset failure is a blow to their self-esteem, yet to a person with a Growth Mindset, it’s an opportunity to improve and find new ways to overcome their obstacles. Same failure, yet different responses. Who is right and who is wrong? Neither. Each person has a different mindset that decides their outcome. Those who are resilient stay in the game longer and draw on their inner means to succeed.
I’ve coached mummy and mom clients who gave up after many years toiling away at their respective goal or dream. It was at that point their biggest breakthrough came. Perhaps all those years of perseverance finally paid off. It was the 19th Century’s minister Henry Ward Beecher who once said: “One’s best success comes after their greatest disappointments.” No one knows what the future holds, so your only guide is whether you can endure repeated defeats and disappointments and still pursue your dream. Consider the advice from the American academic and psychologist Angela Duckworth who writes in Grit: The Power of Passion and Perseverance: “Many of us, it seems, quit what we start far too early and far too often. Even more than the effort a gritty person puts in on a single day, what matters is that they wake up the next day, and the next, ready to get on that treadmill and keep going.”
I know one thing for certain: don’t settle for less than what you’re capable of, but strive for something bigger. Some of you reading this might identify with this message because it resonates with you on a deeper level. For others, at the end of their tether the message might be nothing more than a trivial pep talk. What I wish to convey irrespective of where you are in your journey is: NEVER settle for less. If you settle for less, you will receive less than you deserve and convince yourself you are justified to receive it.
“Two people on a precipice over Yosemite Valley” by Nathan Shipps on Unsplash
Develop A Powerful Vision Of What You Want
“Your problem is to bridge the gap which exists between where you are now and the goal you intend to reach.” — Earl Nightingale
I recall a passage my father often used growing up in 1990s: “Don’t tell me your problems unless you’ve spent weeks trying to solve them yourself.” That advice has echoed in my mind for decades and became my motivator. Don’t leave it to other people or outside circumstances to motivate you because you will be let down every time. It must come from within you. Gnaw away at your problems until you solve them or find a solution. Problems are not stop signs, they are advising you that more work is required to overcome them. Most times, problems help you gain a skill or develop the resources to succeed later. So embrace your challenges and develop the grit to push past them instead of retreat in resignation. Where are you settling in your life right now? Could you be you playing for bigger stakes than you are? Are you willing to play bigger even if it means repeated failures and setbacks? You should ask yourself these questions to decide whether you’re willing to put yourself on the line or settle for less. And that’s fine if you’re content to receive less, as long as you’re not regretful later.
If you have not achieved the success you deserve and are considering giving up, will you regret it in a few years or decades from now? Only you can answer that, but you should carve out time to discover your motivation for pursuing your goals. It’s a fact, if you don’t know what you want you’ll get what life hands you and it may not be in your best interest, affirms author Larry Weidel: “Winners know that if you don’t figure out what you want, you’ll get whatever life hands you.” The key is to develop a powerful vision of what you want and hold that image in your mind. Nurture it daily and give it life by taking purposeful action towards it.
Vision + desire + dedication + patience + daily action leads to astonishing success. Are you willing to commit to this way of life or jump ship at the first sign of failure? I’m amused when I read questions written by millennials on Quora who ask how they can become rich and famous or the next Elon Musk. Success is a fickle and long game with highs and lows. Similarly, there are no assurances even if you’re an overnight sensation, to sustain it for long, particularly if you don’t have the mental and emotional means to endure it. This means you must rely on the one true constant in your favour: your personal development. The more you grow, the more you gain in terms of financial resources, status, success — simple. If you leave it to outside conditions to dictate your circumstances, you are rolling the dice on your future.
So become intentional on what you want out of life. Commit to it. Nurture your dreams. Focus on your development and if you want to give up, know what’s involved before you take the plunge. Because I assure you, someone out there right now is working harder than you, reading more books, sleeping less and sacrificing all they have to realise their dreams and it may contest with yours. Don’t leave your dreams to chance.
'''

## Tokenizing Sentences


In [0]:
sentences=nltk.sent_tokenize(text)

In [11]:
sentences

['Those Who Are Resilient Stay In The Game Longer\n“On the mountains of truth you can never climb in vain: either you will reach a point higher up today, or you will be training your powers so that you will be able to climb higher tomorrow.”\u200a—\u200aFriedrich Nietzsche\nChallenges and setbacks are not meant to defeat you, but promote you.',
 'However, I realise after many years of defeats, it can crush your spirit and it is easier to give up than risk further setbacks and disappointments.',
 'Have you experienced this before?',
 'To be honest, I don’t have the answers.',
 'I can’t tell you what the right course of action is; only you will know.',
 'However, it’s important not to be discouraged by failure when pursuing a goal or a dream, since failure itself means different things to different people.',
 'To a person with a Fixed Mindset failure is a blow to their self-esteem, yet to a person with a Growth Mindset, it’s an opportunity to improve and find new ways to overcome their o

In [0]:
def clean_text(a):
  text=re.sub('[^a-zA-Z]',' ',a)
  text=text.lower()
  return text

## Couting Number of Sentences and Total no of words in each sentence

In [0]:
i=0
doc_info=[]
for sent in sentences:
  sent=clean_text(sent)
  words=nltk.word_tokenize(sent)
  count=0
  for word in words:
    count=count+1
  temp={'doc_id': i,'doc_length': count}
  i=i+1
  doc_info.append(temp)

In [14]:
doc_info

[{'doc_id': 0, 'doc_length': 60},
 {'doc_id': 1, 'doc_length': 26},
 {'doc_id': 2, 'doc_length': 5},
 {'doc_id': 3, 'doc_length': 9},
 {'doc_id': 4, 'doc_length': 16},
 {'doc_id': 5, 'doc_length': 26},
 {'doc_id': 6, 'doc_length': 37},
 {'doc_id': 7, 'doc_length': 5},
 {'doc_id': 8, 'doc_length': 7},
 {'doc_id': 9, 'doc_length': 1},
 {'doc_id': 10, 'doc_length': 10},
 {'doc_id': 11, 'doc_length': 17},
 {'doc_id': 12, 'doc_length': 21},
 {'doc_id': 13, 'doc_length': 9},
 {'doc_id': 14, 'doc_length': 9},
 {'doc_id': 15, 'doc_length': 47},
 {'doc_id': 16, 'doc_length': 37},
 {'doc_id': 17, 'doc_length': 58},
 {'doc_id': 18, 'doc_length': 19},
 {'doc_id': 19, 'doc_length': 19},
 {'doc_id': 20, 'doc_length': 18},
 {'doc_id': 21, 'doc_length': 21},
 {'doc_id': 22, 'doc_length': 85},
 {'doc_id': 23, 'doc_length': 21},
 {'doc_id': 24, 'doc_length': 6},
 {'doc_id': 25, 'doc_length': 13},
 {'doc_id': 26, 'doc_length': 17},
 {'doc_id': 27, 'doc_length': 15},
 {'doc_id': 28, 'doc_length': 17},
 {'

## Frequency of each word in sentence

In [0]:
i=0
frequency_dict_list=[]
for sent in sentences:
  sent=clean_text(sent)
  words=nltk.word_tokenize(sent)
  frequency_dict={}
  for word in words:
    if word in frequency_dict:
      frequency_dict[word]=frequency_dict[word]+1
    else:
      frequency_dict[word]=1
  frequency_dict={'doc_id':i,'frequency_dict':frequency_dict}
  i=i+1
  frequency_dict_list.append(frequency_dict)

In [16]:
frequency_dict_list

[{'doc_id': 0,
  'frequency_dict': {'a': 1,
   'able': 1,
   'and': 1,
   'are': 2,
   'be': 2,
   'but': 1,
   'can': 1,
   'challenges': 1,
   'climb': 2,
   'defeat': 1,
   'either': 1,
   'friedrich': 1,
   'game': 1,
   'higher': 2,
   'in': 2,
   'longer': 1,
   'meant': 1,
   'mountains': 1,
   'never': 1,
   'nietzsche': 1,
   'not': 1,
   'of': 1,
   'on': 1,
   'or': 1,
   'point': 1,
   'powers': 1,
   'promote': 1,
   'reach': 1,
   'resilient': 1,
   'setbacks': 1,
   'so': 1,
   'stay': 1,
   'that': 1,
   'the': 2,
   'those': 1,
   'to': 2,
   'today': 1,
   'tomorrow': 1,
   'training': 1,
   'truth': 1,
   'up': 1,
   'vain': 1,
   'who': 1,
   'will': 3,
   'you': 6,
   'your': 1}},
 {'doc_id': 1,
  'frequency_dict': {'after': 1,
   'and': 2,
   'can': 1,
   'crush': 1,
   'defeats': 1,
   'disappointments': 1,
   'easier': 1,
   'further': 1,
   'give': 1,
   'however': 1,
   'i': 1,
   'is': 1,
   'it': 2,
   'many': 1,
   'of': 1,
   'realise': 1,
   'risk': 1,
  

## Calculating Tf {Term Frequency}
#### Tf= (frequency of word in the sentence / total number of words in the sentence)


In [0]:
tf_score=[]
for temp_dict in frequency_dict_list:
  uid=temp_dict['doc_id']
  for k in temp_dict['frequency_dict']:
    temp={'doc_id':uid,'Tf_Score':temp_dict['frequency_dict'][k]/doc_info[uid]['doc_length'],'key':k}
    tf_score.append(temp)

In [18]:
tf_score

[{'Tf_Score': 0.016666666666666666, 'doc_id': 0, 'key': 'those'},
 {'Tf_Score': 0.016666666666666666, 'doc_id': 0, 'key': 'who'},
 {'Tf_Score': 0.03333333333333333, 'doc_id': 0, 'key': 'are'},
 {'Tf_Score': 0.016666666666666666, 'doc_id': 0, 'key': 'resilient'},
 {'Tf_Score': 0.016666666666666666, 'doc_id': 0, 'key': 'stay'},
 {'Tf_Score': 0.03333333333333333, 'doc_id': 0, 'key': 'in'},
 {'Tf_Score': 0.03333333333333333, 'doc_id': 0, 'key': 'the'},
 {'Tf_Score': 0.016666666666666666, 'doc_id': 0, 'key': 'game'},
 {'Tf_Score': 0.016666666666666666, 'doc_id': 0, 'key': 'longer'},
 {'Tf_Score': 0.016666666666666666, 'doc_id': 0, 'key': 'on'},
 {'Tf_Score': 0.016666666666666666, 'doc_id': 0, 'key': 'mountains'},
 {'Tf_Score': 0.016666666666666666, 'doc_id': 0, 'key': 'of'},
 {'Tf_Score': 0.016666666666666666, 'doc_id': 0, 'key': 'truth'},
 {'Tf_Score': 0.1, 'doc_id': 0, 'key': 'you'},
 {'Tf_Score': 0.016666666666666666, 'doc_id': 0, 'key': 'can'},
 {'Tf_Score': 0.016666666666666666, 'doc_i

## Calculating Idf {Inverse Document Frequency}
#### idf= log( no of sentences / number of sentences containing the word)

In [0]:
idf_score=[]
for temp_dict in frequency_dict_list:
  uid=temp_dict['doc_id']
  for k in temp_dict['frequency_dict']:
    count=sum(k in temp_dict['frequency_dict'] for temp_dict in frequency_dict_list)
    temp={'doc_id':uid,'Idf_Score':math.log(len(doc_info)/count),'key':k}
    idf_score.append(temp)

In [20]:
idf_score

[{'Idf_Score': 2.8526314299133175, 'doc_id': 0, 'key': 'those'},
 {'Idf_Score': 2.005333569526114, 'doc_id': 0, 'key': 'who'},
 {'Idf_Score': 1.3862943611198906, 'doc_id': 0, 'key': 'are'},
 {'Idf_Score': 3.258096538021482, 'doc_id': 0, 'key': 'resilient'},
 {'Idf_Score': 3.258096538021482, 'doc_id': 0, 'key': 'stay'},
 {'Idf_Score': 1.466337068793427, 'doc_id': 0, 'key': 'in'},
 {'Idf_Score': 0.9067212808580044, 'doc_id': 0, 'key': 'the'},
 {'Idf_Score': 2.8526314299133175, 'doc_id': 0, 'key': 'game'},
 {'Idf_Score': 3.258096538021482, 'doc_id': 0, 'key': 'longer'},
 {'Idf_Score': 1.5533484457830569, 'doc_id': 0, 'key': 'on'},
 {'Idf_Score': 3.9512437185814275, 'doc_id': 0, 'key': 'mountains'},
 {'Idf_Score': 1.2431935174792172, 'doc_id': 0, 'key': 'of'},
 {'Idf_Score': 3.9512437185814275, 'doc_id': 0, 'key': 'truth'},
 {'Idf_Score': 0.550046336919272, 'doc_id': 0, 'key': 'you'},
 {'Idf_Score': 2.159484249353372, 'doc_id': 0, 'key': 'can'},
 {'Idf_Score': 3.258096538021482, 'doc_id': 

## Calculating TF-IDF value
#### tfidf= tf-value * idf-value

In [0]:
tfidf_score=[]
for i in tf_score:
  for j in idf_score:
    if j['key']==i['key'] and j['doc_id']==i['doc_id']:
      value=i['Tf_Score']*j['Idf_Score']
      temp={'doc_id':j['doc_id'],'TfIdf_Score':value,'key':j['key']}
      break
  tfidf_score.append(temp)

In [22]:
tfidf_score

[{'TfIdf_Score': 0.047543857165221956, 'doc_id': 0, 'key': 'those'},
 {'TfIdf_Score': 0.03342222615876857, 'doc_id': 0, 'key': 'who'},
 {'TfIdf_Score': 0.046209812037329684, 'doc_id': 0, 'key': 'are'},
 {'TfIdf_Score': 0.054301608967024705, 'doc_id': 0, 'key': 'resilient'},
 {'TfIdf_Score': 0.054301608967024705, 'doc_id': 0, 'key': 'stay'},
 {'TfIdf_Score': 0.04887790229311423, 'doc_id': 0, 'key': 'in'},
 {'TfIdf_Score': 0.030224042695266812, 'doc_id': 0, 'key': 'the'},
 {'TfIdf_Score': 0.047543857165221956, 'doc_id': 0, 'key': 'game'},
 {'TfIdf_Score': 0.054301608967024705, 'doc_id': 0, 'key': 'longer'},
 {'TfIdf_Score': 0.025889140763050947, 'doc_id': 0, 'key': 'on'},
 {'TfIdf_Score': 0.06585406197635713, 'doc_id': 0, 'key': 'mountains'},
 {'TfIdf_Score': 0.020719891957986954, 'doc_id': 0, 'key': 'of'},
 {'TfIdf_Score': 0.06585406197635713, 'doc_id': 0, 'key': 'truth'},
 {'TfIdf_Score': 0.0550046336919272, 'doc_id': 0, 'key': 'you'},
 {'TfIdf_Score': 0.035991404155889535, 'doc_id': 0

## Calculating Each Sentence Score

In [0]:
sentence_score=[]
for doc in doc_info:
  score=0
  uid=doc['doc_id']
  for s in tfidf_score:
    if s['doc_id']==doc['doc_id']:
      score=score + s['TfIdf_Score']
  temp={'doc_id':doc['doc_id'],'sent_score':score,'sentence':sentences[uid]}
  sentence_score.append(temp)

In [24]:
sentence_score

[{'doc_id': 0,
  'sent_score': 2.4848160665852275,
  'sentence': 'Those Who Are Resilient Stay In The Game Longer\n“On the mountains of truth you can never climb in vain: either you will reach a point higher up today, or you will be training your powers so that you will be able to climb higher tomorrow.”\u200a—\u200aFriedrich Nietzsche\nChallenges and setbacks are not meant to defeat you, but promote you.'},
 {'doc_id': 1,
  'sent_score': 2.434333786143525,
  'sentence': 'However, I realise after many years of defeats, it can crush your spirit and it is easier to give up than risk further setbacks and disappointments.'},
 {'doc_id': 2,
  'sent_score': 2.5332283514262093,
  'sentence': 'Have you experienced this before?'},
 {'doc_id': 3,
  'sent_score': 2.1603233148827012,
  'sentence': 'To be honest, I don’t have the answers.'},
 {'doc_id': 4,
  'sent_score': 2.040123555450907,
  'sentence': 'I can’t tell you what the right course of action is; only you will know.'},
 {'doc_id': 5,
  '

## Calculating The Avg Sentence Score and Standard Deviation

In [0]:
sum_score=0
array=[]
for sent in sentence_score:
  sum_score=sum_score+sent['sent_score']
  array.append(sent['sent_score'])

In [0]:
avg=sum_score/len(sentence_score)

In [0]:
stdev=statistics.stdev(array)

In [28]:
avg

2.5134703761868433

In [29]:
stdev

0.38044568209836954

## Summarizing The Text by selecting the sentence greater than avg sentence score

In [0]:
summary_sent=[]
for i in sentence_score:
  if i['sent_score']>=avg+1*stdev:
    summary_sent.append(i['sentence'])

In [0]:
summary=' '.join(summary_sent)

In [41]:
summary

'Same failure, yet different responses. Neither. Each person has a different mindset that decides their outcome. Perhaps all those years of perseverance finally paid off. Consider the advice from the American academic and psychologist Angela Duckworth who writes in Grit: The Power of Passion and Perseverance: “Many of us, it seems, quit what we start far too early and far too often. Vision + desire + dedication + patience + daily action leads to astonishing success. I’m amused when I read questions written by millennials on Quora who ask how they can become rich and famous or the next Elon Musk.'

## Summarized Sentence

'Same failure, yet different responses. Neither. Each person has a different mindset that decides their outcome. Perhaps all those years of perseverance finally paid off. Consider the advice from the American academic and psychologist Angela Duckworth who writes in Grit: The Power of Passion and Perseverance: “Many of us, it seems, quit what we start far too early and far too often. Vision + desire + dedication + patience + daily action leads to astonishing success. I’m amused when I read questions written by millennials on Quora who ask how they can become rich and famous or the next Elon Musk.'

##### To make the summary more short and for large doumments it threshold can be selected as (avg + 1.5 * stdev) or (avg + 3 * stdev)